### DISCUSSION TITLE
"multiple snapshots cannot be sent from a bookmark"

### DISCUSSION THREAD

**Discussion by sgsunder** [openzfs/zfs#15856] [Q&A] [answered]
I'm receiving the following error:
```
Error: multiple snapshots cannot be sent from a bookmark.
```

when trying to run the following command:
```shell
sudo zfs send -wR -i "tank/service/gitea#backup-2024-01-14" "tank/service/gitea@backup-2024-02-06" \
    | pv | ssh backup zfs receive -Fu "backup/gitea"
```

I'm not sure why, as I only have the target snapshot present, so there shouldn't be multiple snapshots being sent:
```shell
$ zfs list -t snapshot tank/service/gitea
NAME                                                        USED  AVAIL  REFER  MOUNTPOINT
tank/service/gitea@backup-2024-02-06                          0B      -  1.28G  -

$ zfs list -t bookmark tank/service/gitea
NAME                                   USED  AVAIL  REFER  MOUNTPOINT
tank/service/gitea#backup-2024-01-14      -      -  1.27G  -
tank/service/gitea#backup-2024-02-06      -      -  1.28G  -
```

Other info:
- I'm sending it as a raw, encrypted snapshot. I want it to stay encrypted on the backup target.
- Source machine is Debian 12, running `zfs-2.2.2-4~bpo12+1` and `zfs-kmod-2.2.2-4~bpo12+1`

Any idea what I am doing wrong?

**Comment by rincebrain** [comment, marked as answer]
It says in the message that you cannot do that, and why.

The man page says:
```
     zfs send [-DLPVRbcehnpsvw] [[-I|-i] snapshot] snapshot
     zfs send [-DLPVcensvw] [-i snapshot|bookmark] filesystem|volume|snapshot
```

note how -R and -I are not included in the one that can take bookmarks.

Conceptually, one could implement support for -I or -R with bookmarks, but nobody has. Feel free to doso if you need it, otherwise I'd suggest just sending one and then -R -i from that one to the rest, if you need it.

(I'd also generally discourage using -R, because it tries to do "all in one" but it turns out there's a lot of really gross edge cases that you can't handle automatically, so you'd end up needing a lot of tooling to very complicatedly handle cleanup if anything goes wrong...)

**Comment by sgsunder** [comment]
Would -Lec be a suitable replacement for -R if I don’t care about intermediate snapshots being sent and don’t have any child datasets?

(I would like to add -b as well but that doesn’t seem to be possible with bookmarks it seems. Not critical for me but would be nice)

**Reply by rincebrain** [reply]
-Lec is implied by -w.
